File: js/text-processor.js
/**
 * TextProcessor - Converts text input to viseme sequences.
 * Uses rule-based phoneme approximation (no dictionary needed).
 *
 * Viseme Set (12 shapes):
 *  0: Neutral - Rest/silence, mouth closed
 *  1: Aa      - Wide open "ah" (father, cat)
 *  2: D       - Tongue on ridge (d, t, n)
 *  3: Ee      - Wide smile (ee, ih — "see", "bit")
 *  4: F       - Lower lip under upper teeth (f, v)
 *  5: L       - Tongue tip up (l)
 *  6: M       - Lips pressed (m, b, p)
 *  7: Oh      - Rounded open (oh, aw — "boat", "saw")
 *  8: R       - Slight pucker/open (r)
 *  9: S       - Teeth together narrow (s, z, sh, ch)
 * 10: Uh      - Slightly open relaxed (uh, schwa)
 * 11: W-Oo    - Pursed/rounded (w, oo — "we", "blue")
 */
class TextProcessor {
  constructor() {
    this.visemeQueue = [];
    this.defaultDuration = 100; // ms per phoneme
    this.minDuration = 80;
    this.maxDuration = 150;
    this.visemeCount = 12; // Total number of viseme shapes (0-11)

    // Digraph mappings (checked first)
    this.digraphMap = {
      'th': 2,   // D — tongue on ridge
      'sh': 9,   // S — teeth narrow
      'ch': 9,   // S — teeth narrow
      'zh': 9,   // S — teeth narrow
      'ph': 4,   // F — lip under teeth
      'wh': 11,  // W-Oo — pursed
      'oo': 11,  // W-Oo — pursed rounded
      'ee': 3,   // Ee — wide smile
      'ea': 3,   // Ee — wide smile
      'ai': 1,   // Aa — wide open
      'ay': 1,   // Aa — wide open
      'oa': 7,   // Oh — rounded open
      'ou': 7,   // Oh — rounded open
      'ow': 7,   // Oh — rounded open
      'oi': 7,   // Oh — rounded open
      'au': 1    // Aa — wide open
    };

    // Single character mappings
    this.charMap = {
      'a': 1,  'e': 3,  'i': 3,  'o': 7,  'u': 10,
      'p': 6,  'b': 6,  'm': 6,
      'f': 4,  'v': 4,
      's': 9,  'z': 9,  'j': 9,
      't': 2,  'd': 2,  'n': 2,
      'l': 5,  'r': 8,
      'k': 2,  'g': 2,  'c': 9,
      'w': 11, 'y': 3,
      'h': 10, 'q': 6,  'x': 9
    };
  }

  /**
   * Normalize text: lowercase, remove non-letter characters.
   * @param {string} text
   * @returns {string}
   */
  normalize(text) {
    if (!text || typeof text !== 'string') return '';
    return text.toLowerCase().replace(/[^a-z\s]/g, '');
  }

  /**
   * Map a single character or digraph to a viseme ID.
   * @param {string} chars - 1-2 characters
   * @returns {number} viseme ID (0-11)
   */
  mapToViseme(chars) {
    if (!chars || chars.length === 0) return 0;
    const lower = chars.toLowerCase();
    if (lower.length >= 2 && this.digraphMap[lower.substring(0, 2)] !== undefined) {
      return this.digraphMap[lower.substring(0, 2)];
    }
    return this.charMap[lower[0]] !== undefined ? this.charMap[lower[0]] : 0;
  }

  /**
   * Process a text string into a viseme sequence with durations.
   * @param {string} text - Input text
   * @param {number} [timestamp] - Timestamp when text was received
   * @returns {Array<{viseme: number, duration: number, timestamp: number}>}
   */
  processText(text, timestamp) {
    const ts = timestamp || (typeof performance !== 'undefined' ? performance.now() : Date.now());
    const normalized = this.normalize(text);
    if (!normalized) return [];

    const sequence = [];
    const words = normalized.split(/\s+/).filter(w => w.length > 0);
    let currentTime = ts;

    for (const word of words) {
      let i = 0;
      while (i < word.length) {
        let viseme;
        let consumed = 1;

        // Check digraphs first
        if (i + 1 < word.length) {
          const digraph = word.substring(i, i + 2);
          if (this.digraphMap[digraph] !== undefined) {
            viseme = this.digraphMap[digraph];
            consumed = 2;
          }
        }

        if (viseme === undefined) {
          viseme = this.charMap[word[i]] !== undefined ? this.charMap[word[i]] : 0;
        }

        const duration = this._computeDuration(word, i);
        sequence.push({
          viseme: viseme,
          duration: duration,
          timestamp: currentTime
        });
        currentTime += duration;
        i += consumed;
      }

      // Add a brief rest between words
      sequence.push({
        viseme: 0,
        duration: 50,
        timestamp: currentTime
      });
      currentTime += 50;
    }

    return sequence;
  }

  /**
   * Compute duration for a phoneme based on position in word.
   * Vowels get slightly longer duration.
   */
  _computeDuration(word, index) {
    const char = word[index];
    const isVowel = 'aeiou'.includes(char);
    const base = isVowel ? 120 : 90;
    return Math.max(this.minDuration, Math.min(this.maxDuration, base));
  }

  /**
   * Add a text chunk to the viseme queue for incremental processing.
   * @param {string} text
   */
  addTextChunk(text) {
    const sequence = this.processText(text);
    this.visemeQueue.push(...sequence);
  }

  /**
   * Get the next viseme from the queue.
   * @returns {{viseme: number, duration: number, timestamp: number}|null}
   */
  getNextViseme() {
    if (this.visemeQueue.length === 0) return null;
    return this.visemeQueue.shift();
  }

  /**
   * Peek at the next viseme without removing it.
   */
  peekNextViseme() {
    if (this.visemeQueue.length === 0) return null;
    return this.visemeQueue[0];
  }

  /**
   * Get remaining queue length.
   */
  getQueueLength() {
    return this.visemeQueue.length;
  }

  /**
   * Clear the viseme queue.
   */
  clearQueue() {
    this.visemeQueue = [];
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = TextProcessor;
}

File: text_processor.py
"""
Text-to-viseme processor.

Converts text input into a sequence of viseme indices with estimated durations.
Uses rule-based phoneme approximation (digraphs, vowels, consonants).

12-viseme system:
  0: Neutral   1: Aa   2: D    3: Ee   4: F    5: L
  6: M         7: Oh   8: R    9: S   10: Uh  11: W-Oo
"""

from typing import List, Dict


class VisemeEvent:
    __slots__ = ("viseme", "duration", "char_source")

    def __init__(self, viseme: int, duration: float, char_source: str = ""):
        self.viseme = viseme
        self.duration = duration
        self.char_source = char_source

    def to_dict(self) -> dict:
        return {
            "viseme": self.viseme,
            "duration": self.duration,
            "char_source": self.char_source,
        }


# Digraph mappings (checked before single chars)
DIGRAPH_MAP: Dict[str, int] = {
    "th": 2,   # D (tongue ridge, th-sound)
    "sh": 9,   # S (teeth narrow)
    "ch": 9,   # S
    "zh": 9,   # S
    "ph": 4,   # F
    "wh": 11,  # W-Oo
    "oo": 11,  # W-Oo
    "ee": 3,   # Ee
    "ea": 3,   # Ee
    "ai": 1,   # Aa
    "ay": 1,   # Aa
    "oa": 7,   # Oh
    "ou": 7,   # Oh
    "ow": 7,   # Oh
    "oi": 7,   # Oh
    "au": 1,   # Aa
}

# Single character mappings
CHAR_MAP: Dict[str, int] = {
    # Vowels
    "a": 1,   # Aa
    "e": 3,   # Ee
    "i": 3,   # Ee
    "o": 7,   # Oh
    "u": 10,  # Uh
    # Bilabials (lips pressed)
    "p": 6,   # M
    "b": 6,   # M
    "m": 6,   # M
    # Labiodentals (lip under teeth)
    "f": 4,   # F
    "v": 4,   # F
    # Sibilants (teeth narrow)
    "s": 9,   # S
    "z": 9,   # S
    "j": 9,   # S
    # Alveolars (tongue on ridge)
    "t": 2,   # D
    "d": 2,   # D
    "n": 2,   # D
    # Lateral/Rhotic
    "l": 5,   # L
    "r": 8,   # R
    # Velars (mapped to tongue ridge)
    "k": 2,   # D
    "g": 2,   # D
    "c": 9,   # S (soft c default)
    # Glides
    "w": 11,  # W-Oo
    "y": 3,   # Ee
    # Others
    "h": 10,  # Uh (breathy)
    "q": 6,   # M
    "x": 9,   # S
}

# Duration constants (milliseconds)
BASE_DURATION = 100.0
MIN_DURATION = 50.0
MAX_DURATION = 200.0
REST_DURATION = 80.0
VOWEL_MULTIPLIER = 1.3
CONSONANT_MULTIPLIER = 0.8


class TextProcessor:
    def __init__(self):
        self.viseme_count = 12
        self.speed_multiplier = 1.0

    def process_text(self, text: str) -> List[dict]:
        """
        Convert text to a sequence of viseme events with durations.

        Args:
            text: Input text string

        Returns:
            List of viseme event dicts with viseme index, duration, and source char
        """
        if not text or not text.strip():
            return []

        text = text.lower().strip()
        events: List[VisemeEvent] = []
        i = 0

        while i < len(text):
            ch = text[i]

            # Whitespace -> rest viseme
            if ch == " ":
                # Merge consecutive spaces
                if not events or events[-1].viseme != 0:
                    events.append(VisemeEvent(0, REST_DURATION, " "))
                i += 1
                continue

            # Skip non-alpha
            if not ch.isalpha():
                i += 1
                continue

            # Check digraphs first
            if i + 1 < len(text):
                digraph = text[i : i + 2]
                if digraph in DIGRAPH_MAP:
                    viseme = DIGRAPH_MAP[digraph]
                    dur = self._compute_duration(digraph, is_vowel=self._is_vowel_digraph(digraph))
                    events.append(VisemeEvent(viseme, dur, digraph))
                    i += 2
                    continue

            # Single character
            if ch in CHAR_MAP:
                viseme = CHAR_MAP[ch]
                dur = self._compute_duration(ch, is_vowel=ch in "aeiou")
                events.append(VisemeEvent(viseme, dur, ch))
            # Unknown char -> skip
            i += 1

        return [e.to_dict() for e in events]

    def _compute_duration(self, chars: str, is_vowel: bool) -> float:
        """Compute duration for a phoneme unit."""
        base = BASE_DURATION
        if is_vowel:
            base *= VOWEL_MULTIPLIER
        else:
            base *= CONSONANT_MULTIPLIER

        base /= self.speed_multiplier
        return max(MIN_DURATION, min(MAX_DURATION, base))

    def _is_vowel_digraph(self, digraph: str) -> bool:
        """Check if a digraph represents a vowel sound."""
        vowel_digraphs = {"oo", "ee", "ea", "ai", "ay", "oa", "ou", "ow", "oi", "au"}
        return digraph in vowel_digraphs

    def set_speed(self, multiplier: float):
        """Set speed multiplier (0.5 = half speed, 2.0 = double speed)."""
        self.speed_multiplier = max(0.25, min(4.0, multiplier))
